server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java

```diff
@@ -114,19 +114,36 @@ public synchronized CompletableFuture<CobolDocumentModel> scheduleAnalysis(Strin
    * @param eventSource source of the event
    * @return document model with analysis result
    */
-  public synchronized CompletableFuture<CobolDocumentModel> scheduleAnalysis(String uri, String text, Integer currentRevision, boolean open, boolean force, SourceUnitGraph.EventSource eventSource) {
+  public synchronized FutureTask<CobolDocumentModel> scheduleAnalysis(String uri, String text, Integer currentRevision, boolean open, boolean force, SourceUnitGraph.EventSource eventSource) {
     notifyAllListeners(AnalysisState.SCHEDULED, documentModelService.get(uri), eventSource);
     String id = makeId(uri, currentRevision);
     Integer prevId = analysisResultsRevisions.put(uri, currentRevision);
     if (currentRevision.equals(prevId) && !force) {
       notifyAllListeners(AnalysisState.SKIPPED, documentModelService.get(uri), eventSource);
       return analysisResults.get(id);
     }
-    Executor analysisExecutor = getExecutor(uri);
-    CompletableFuture<CobolDocumentModel> value = CompletableFuture.supplyAsync(() -> {
+    ExecutorService analysisExecutor = getExecutor(uri);
+    CobolDocumentModel documentModel = documentModelService.get(uri);
+    documentModel.updateLastAnalysisNotCancelled();
+    FutureTask<CobolDocumentModel> futureTask = new FutureTask<>(scheduleAnalysis(uri, documentModel, text, currentRevision, open, force, eventSource, id));
+    analysisResults.put(id, futureTask);
+    analysisExecutor.submit(futureTask);
+    if (prevId != null && !force) {
+      Optional.ofNullable(analysisResults.get(makeId(uri, prevId))).ifPresent(cf -> cf.cancel(true));
+    }
```

Comment:
Shouldn't this be in a similar loop that can be found in `reanalyseOpenedPrograms`?

Reply:
I am under impression that this method is responsible to scheduling an analysis and the caller of this method should try to cancel previous analysis before calling this method.
But I think, `reanalyseCopybooksAssociatedPrograms` should be updated

Reply:
Fine with me, but be also aware of the other direct calls to the `scheduleAnalysis` function - from didOpen, didChange handlers and AnalysisHandler.getDependencies.

Reply:
to be handled in #2467.